modin/core/dataframe/pandas/partitioning/partition_manager.py

```diff
@@ -1285,3 +1285,103 @@ def finalize(cls, partitions):
+
+        Returns
+        -------
+        tuple
+            Tuple of (new_partitions, new_lengths, new_widths)
+        """
+
+        def compute_schema(axis_lengths, axis):
+            """
+            Compute repartition schema.
+
+            The schema is a sequence of intervals in which the partitions to be combined are defined.
+            The interval is determined by the chunksize.
+
+            Parameters
+            ----------
+            axis_lengths : list
+                The length/width of each partition in the rows/columns.
+            axis : {0, 1}
+                Axis for creating the repartition scheme.
+
+            Returns
+            -------
+            tuple
+                Tuple of (schema, updated axis lengths aсcording to the schema)
+            """
+            from modin.core.storage_formats.pandas.utils import (
+                compute_chunksize,
+            )
+
+            idx = pandas.RangeIndex(sum(axis_lengths))
+            df = pandas.DataFrame(**{"columns" if axis == 1 else "index": idx})
+            chunksize = compute_chunksize(df, NPartitions.get(), axis=axis)
+
+            schema = []
+            # initialize the first interval
+            schema.append([0, None])
+            current_axis_length = 0
+
+            for idx, axis_length in enumerate(axis_lengths):
+                if current_axis_length + axis_length < chunksize:
+                    current_axis_length += axis_length
+                    continue
+                current_axis_length = 0
+                # update end of the interval
+                schema[-1][1] = idx + 1
+                if idx + 1 != len(axis_lengths):
+                    schema.append([idx + 1, None])
+            if schema[-1][1] is None:
+                schema[-1][1] = len(axis_lengths)
+
+            # update axis lengths aсcording to the schema
+            new_axis_lengths = list(
+                map(lambda t: sum(axis_lengths[t[0] : t[1]]), schema)
+            )
+            return schema, new_axis_lengths
+
+        axis_lengths = widths if axis == 1 else lengths
+        repartition_schema, axis_lengths = compute_schema(axis_lengths, axis)
+
+        new_lengths = axis_lengths if axis == 0 else lengths
+        new_widths = axis_lengths if axis == 1 else widths
+
+        new_partitions = []
+        for start, end in repartition_schema:
+            _parts = partitions[:, start:end] if axis == 1 else partitions[start:end]
+            if end - start == 1:
+                # Fast path; there is only one column/row partition in the merge interval
+                new_partitions.append(_parts)
+                continue
+            new_partitions.append(
+                cls.map_axis_partitions(
+                    axis=axis,
+                    partitions=_parts,
+                    map_func=lambda df: df,
+                    keep_partitioning=False,
+                    lengths=[sum(new_widths) if axis == 1 else sum(new_lengths)],
```

Comment:
```suggestion
                    lengths=[sum(axis_lengths)],
```
but I'm in doubt about this piece - why `lengths` is identical for each `map_axis_partitions()` call regardless of `start` and `end`?..

Reply:
the length of the whole axis is used here, so that the reverse repartitioning does not occur, since it is necessary that the passed length be greater than the sum of the merged partitions lengths.

Reply:
The comment is added.